### PR TITLE
8328146: Set LIBCXX automatically

### PR DESCRIPTION
**Pull request by magicus**
We are adding LIBCXX to LIBS in calls to SetupJdkLibrary whenever LINK_TYPE is C++. We should do this automatically in SetupJdkLibrary for C++ linking.

I also removed the superfluous `-lc` from some places where it had been added.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328146](https://bugs.openjdk.org/browse/JDK-8328146): Set LIBCXX automatically (**Bug** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18298/head:pull/18298` \
`$ git checkout pull/18298`

Update a local copy of the PR: \
`$ git checkout pull/18298` \
`$ git pull https://git.openjdk.org/jdk.git pull/18298/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18298`

View PR using the GUI difftool: \
`$ git pr show -t 18298`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18298.diff">https://git.openjdk.org/jdk/pull/18298.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18298#issuecomment-1997176180)

### CODE REVIEW

**Comment by magicus** [comment]
I have verified with COMPARE_BUILD that there is no significant difference in the build result with or without this patch on Oracle default CI platforms (linux x64/aarch64, macos x64/aarch64, windows x64).

File: make/modules/java.base/Lib.gmk [hunk 1]
```diff
@@ -213,9 +213,8 @@ $(eval $(call SetupJdkLibrary, BUILD_SYSLOOKUPLIB, \
     LDFLAGS := $(LDFLAGS_JDKLIB), \
     LDFLAGS_linux := -Wl$(COMMA)--no-as-needed, \
     LDFLAGS_aix := -brtl -bexpfull, \
-    LIBS := $(LIBCXX), \
-    LIBS_linux := -lc -lm -ldl, \
-    LIBS_aix := -lc -lm -ldecNumber, \
+    LIBS_linux := -lm -ldl, \
+    LIBS_aix := -lm -ldecNumber, \
```
Comment (erikj79) [inline make/modules/java.base/Lib.gmk:217]: This lib doesn't set LINK_TYPE. Did it add LIBCXX without needing it?

File: make/modules/jdk.jpackage/Lib.gmk [hunk 2]
```diff
@@ -129,7 +129,7 @@ ifeq ($(call isTargetOs, windows), true)
       CXXFLAGS := $(CXXFLAGS_JDKLIB) $(JPACKAGE_CXXFLAGS_windows), \
       LDFLAGS := $(LDFLAGS_JDKLIB) $(LDFLAGS_CXX_JDK) \
           $(call SET_SHARED_LIBRARY_ORIGIN), \
-      LIBS := $(LIBCXX) advapi32.lib ole32.lib msi.lib user32.lib \
+      LIBS := advapi32.lib ole32.lib msi.lib user32.lib \
```
Comment (erikj79) [inline make/modules/jdk.jpackage/Lib.gmk:132]: Should this (and the others below) be using LINK_TYPE:=C++ or was $(LIBCXX) never actually needed on Windows?

File: make/modules/java.base/Lib.gmk (hunk 1 above)
Comment (magicus) [reply make/modules/java.base/Lib.gmk:217]: Apparently. It is a minimal library, consisting of a single, (almost) empty file. It also set `-lc` which is not needed. My guess is that this is the result of copy-paste code.

File: make/modules/jdk.jpackage/Lib.gmk (hunk 2 above)
Comment (magicus) [reply make/modules/jdk.jpackage/Lib.gmk:132]: $(LIBCXX) is not needed on Windows. It is used only on gcc, where it expands to flags needed to link statically with the C++ standard libraries, if static linking with the standard libs is selected.

There's overall a lot of messiness with the flags, with setting flags that is pointless, or omitting to set flags that we should (due to copy/paste once again). These initial PRs I've posted try to keep the same flags for all libraries. The fun (or "fun") part comes after that, when I'll try to modify the flags to make more sense.

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**